phoenix-core-client/src/main/java/org/apache/phoenix/expression/util/bson/UpdateExpressionUtils.java

```diff
@@ -0,0 +1,950 @@
+      if (topLevelValue != null) {
+        BsonValue value = modifyFieldValueByDelete(topLevelValue, newVal);
+        if (value == null) {
+          bsonDocument.remove(fieldKey);
+        } else {
+          bsonDocument.put(fieldKey, value);
+        }
+      } else if (!fieldKey.contains(".") && !fieldKey.contains("[")) {
+        LOGGER.info("Nothing to be removed as field with key {} does not exist", fieldKey);
+      } else {
+        // If the top level field does not exist and the field key contains "." or "[]" notations
+        // for nested document or nested array, use the self-recursive function to go through
+        // the document tree, one level at a time.
+        updateNestedFieldByDelete(bsonDocument, 0, fieldKey, newVal);
+      }
+    }
+  }
+
+  /**
+   * Update the nested field with $DELETE_FROM_SET operation. The field key is expected to contain
+   * "." and/or "[]" notations for nested documents and/or nested array elements. This function
+   * keeps recursively calling itself until it reaches the leaf node in the given tree.
+   * For instance, for field key "category.subcategories.brands[5]", first the function
+   * evaluates and retrieves the value for top-level field "category". The value of "category"
+   * is expected to be nested document. First function call has value as full document, it retries
+   * nested document under "category" field and calls the function recursively with index value
+   * same as index value of first "." (dot) in the field key. For field key
+   * "category.subcategories.brands[5]", the index value would be 8. The second function call
+   * retrieves value of field key "subcategories", which is expected to be nested document.
+   * The third function call gets this nested document as BsonValue and index value as 22 as the
+   * field key has second "." (dot) notation at index 22. The third function call searches for
+   * field key "brands" and expects its value as nested array. The forth function call gets
+   * this nested array as BsonValue and index 29 as the field key has "[]" array notation starting
+   * at index 29. The forth function call retrieves value of nested array element at index 5.
+   * As the function is at leaf node in the tree, now it performs the $DELETE_FROM_SET operation
+   * as per the $DELETE_FROM_SET semantics.
+   *
+   * @param value Bson value at the given level of the document hierarchy.
+   * @param idx The index value for the given field key. The function is expected to retrieve
+   * the value of the nested document or array at the given level of the tree.
+   * @param fieldKey The full field key.
+   * @param setValuesToDelete The set values that need to be removed from the existing set.
+   */
+  private static void updateNestedFieldByDelete(final BsonValue value,
+      final int idx, final String fieldKey, final BsonValue setValuesToDelete) {
+    int curIdx = idx;
+    if (fieldKey.charAt(curIdx) == '.') {
```

Comment:
There is too much repeated code here to find a entry (key-value) in a document. We can reduce the code by having a common method that returns an entry for a given field from an update expression. The same method can be reused for the fields from a condition expression too, I think. By doing so, the recursion and iterating over keys for . and [ can be contained and implemented in one place.

Reply:
For condition expression we already have common utility method that perform tree traversal one level at a time and retrieves the deeply nested fields: `CommonComparisonExpressionUtils#getFieldFromDocument()`. This is being used at many places.
For update expression, it is really complicated to maintain such common tree traversal logic because we not only need to traverse the tree, we also need to update the leaf node or intermediate node of the tree. Doing so requires different BSON APIs for different update operation. I have tried to keep common logic separate but its really making update expression very complicated. Some behaviors are unique to individual update operation e.g. DELETE_FROM_SET needs to remove the field if we remove all elements of the set. ADD needs to perform arithmetic or set union operation depending on the data type. Maintaining all this logic with tree traversal as separate common function is becoming too complex to manage.

Reply:
Let me try refactoring some portion within update expression, however sharing the common portion with condition expression does not seem feasible due to inherent complexity involved with individual update operators.

Reply:
Refactor for update expression done https://github.com/apache/phoenix/pull/1906/commits/34f3675bbb623275e7344c195a2f0e66afdde364

Reply:
Your refactoring reduced the code significantly. However, you did it in a different way than what I suggested. I did not suggested writing a common method for updating a field. As you can see from your refactoring, the code reduction comes from not repeating the code that traverses the BSON document to find the entry for a given field name (the path to an entry). Your recursive methods return void currently. They are written in procedural programming (using side effects). The common (recursive) method I suggested should be written in functional programming (without any side effect). It should take a Bson document and a filed name and return an entry. Does that make sense?

Reply:
While condition expression requires the retrieval of the field name key-value entry by traversing the tree, update expression has complicated behavior depending on the update operation. So if we were to write function that returns BsonValue, it would be for the ancestor node rather than the leaf node that requires update. For example, path1.path2.path3[4] requires function returning path1.path2.path3 so that update expression can perform its logic on path3, whereas for path1.path2.path3, the function needs to return path1.path2 value. While this seems doable, but we need separate function anyway because now it needs to return the ancestor node and not the leaf node itself.

Such function cannot be shared with the condition expression.

Reply:
Also, even if we use function to return ancestor node, for (deeply) nested arrays, the caller still needs the array index on which to perform the update operation so the caller will need to parse the field key string from the last index position to find the leaf node (nested array) index position and then perform the operation. Hence, this approach is also not any simpler than the current way of doing recursion (procedural way) in which while we return void, we do take care of updating the value at the target leaf node. The recursion followed by updating the target node is not expected to be buggy.

Reply:
I did not understand the complication you mentioned if the method I suggested returns the list of Bson documents instead of just the document corresponding to the last component of the path. The Bson docs can be added to the list when the method traverses the entire document from top to bottom. I am not suggesting to change the restructuring that you have just done so far. I am suggesting that you can further improve your restructuring by extracting the find method as a common method. The problem statement begs for this method as it is "update/delete/check the value of the bson document for a given path".

Reply:
The list of BsonValue can be returned but depending on whether the update method (caller) overrides the already present BsonValue or needs to add new BsonValue, it needs to have field key associated with each BsonValue in the list. While we can use LinkedHashMap and let caller get last pair from LinkedHashMap<String, BsonValue>, the caller still needs to traverse the field key again to get the last portion of the key back. For example, for path1.path2.path3[4], the caller still needs to recognize and have different logic for path3[4] vs path4 (if the field key was path1.path2.path3.path4). The caller still needs to retrieve the value of path4 or path3[4] as an additional parsing.

The current version of update recursive function looks far simpler.

Reply:
I did not understand the need for LinkedHashMap instead of just ArrayList and did not understand why doc vs array update cannot be resolved easily. Let's have a side conversation on this.